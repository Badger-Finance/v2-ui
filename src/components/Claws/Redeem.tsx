import React, { FC, useState } from 'react';
import { Box, Button, Grid, InputBase, makeStyles, Typography } from '@material-ui/core';
import ClawParams, { ClawParam } from './ClawParams';
import { useMainStyles } from './index';
import ClawLabel from './ClawLabel';
import ClawDetails from './ClawDetails';

const options = ['eCLAW FEB29', 'eCLAW MARCH29'];

const eCLAWS: Record<string, string> = {
	'eCLAW FEB29': '1000',
	'eCLAW MARCH29': '2000',
};

const redeemableTokens: Record<string, string> = {
	'eCLAW FEB29': 'wBTCwETHSLP',
	'eCLAW MARCH29': 'bBadger',
};

const initialValue = { amount: '0.00' };

const useStyles = makeStyles((theme) => ({
	border: {
		border: '1px solid #5C5C5C',
		borderRadius: 8,
	},
	selectContainer: {
		[theme.breakpoints.only('xs')]: {
			justifyContent: 'space-between',
		},
		[theme.breakpoints.up('lg')]: {
			paddingLeft: '10%',
		},
	},
	margin: {
		margin: theme.spacing(1),
	},
	centered: {
		margin: 'auto',
	},
}));

const Redeem: FC = () => {
	const mainClasses = useMainStyles();
	const classes = useStyles();
	const [redeemParams, setRedeemParams] = useState<ClawParam>(initialValue);

	return (
		<Grid container>
			<Box clone pb={2}>
				<Grid item xs={12}>
					<Box clone pb={1}>
						<Grid item xs={12}>
							<ClawLabel
								name="Token"
								balanceLabel={
									redeemParams.selectedOption && `Available ${redeemParams.selectedOption}:`
								}
								balance={redeemParams.selectedOption && eCLAWS[redeemParams.selectedOption]}
							/>
						</Grid>
					</Box>
					<ClawParams
						referenceBalance={redeemParams.selectedOption && eCLAWS[redeemParams.selectedOption]}
						placeholder="Select Token"
						amount={redeemParams.amount}
						onAmountChange={(amount: string, error?: boolean) => {
							setRedeemParams({
								...redeemParams,
								amount,
								error: error ? `Amount exceeds ${redeemParams.selectedOption} balance` : undefined,
							});
						}}
						selectedOption={redeemParams.selectedOption}
						onOptionChange={(selectedOption: string) => {
							setRedeemParams({
								selectedOption,
								amount: '0.00',
								error: undefined,
							});
						}}
						disabledAmount={!redeemParams.selectedOption}
					/>
				</Grid>
			</Box>
			{redeemParams.selectedOption && (
				<Box clone py={2}>
					<Grid item xs={12} sm={8} className={classes.centered}>
						<Box clone pb={1}>
							<Grid item xs={12}>
								<ClawLabel name="You Receive" />
							</Grid>
						</Box>
						<Box clone py={1} px={2}>
							<Grid container alignContent="center" alignItems="center" className={classes.border}>
								<Grid item xs={12}>
									<Grid container alignItems="center" spacing={2} className={classes.selectContainer}>
										<Grid item xs={12} sm={6}>
											<Typography>{redeemableTokens[redeemParams.selectedOption]}</Typography>
										</Grid>
										<Grid item xs={12} sm={6}>
											<InputBase type="tel" disabled placeholder="0.00" value="" />
										</Grid>
									</Grid>
								</Grid>
							</Grid>
						</Box>
					</Grid>
				</Box>
			)}
			<Grid item xs={12}>
				<Grid container className={mainClasses.details}>
					<ClawDetails
						details={[
							{ 'Expiration Date': 'Feb 29, 2021 8:00 UTC' },
							{ 'Expiration Price': '1 eCLAW FEB29 = .000001 wBTCWethSLP' },
						]}
					/>
				</Grid>
			</Grid>
			<Grid item xs={12}>
				<Grid container>
					<Button
						color="primary"
						variant="contained"
						disabled={!!redeemParams.error || !redeemParams.selectedOption}
						size="large"
						className={mainClasses.button}
					>
						{redeemParams.error ? redeemParams.error : 'REDEEM'}
					</Button>
				</Grid>
			</Grid>
		</Grid>
	);
};

export default Redeem;
