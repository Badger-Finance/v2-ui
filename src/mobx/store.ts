import { RouterStore } from 'mobx-router';
import UiState from './reducers';
import WalletStore from './stores/walletStore';
import ContractsStore from './stores/contractsStore';
import AirdropStore from './stores/airdropStore';
import RebaseStore from './stores/rebaseStore';
import RewardsStore from './stores/rewardsStore';
import IbBTCStore from './stores/ibBTCStore';
import TransactionsStore from './stores/transactionsStore';
import SettStoreV2 from './stores/settStoreV2';
import { NETWORK_LIST } from '../config/constants';
import HoneyPotStore from './stores/honeyPotStore';

export class RootStore {
	public router: RouterStore<RootStore>;
	public wallet: WalletStore;
	public uiState: UiState;
	public contracts: ContractsStore;
	public airdrops: AirdropStore;
	public rebase: RebaseStore;
	public rewards: RewardsStore;
	public ibBTCStore: IbBTCStore;
	public setts: SettStoreV2;
	public transactions: TransactionsStore;
	public honeyPot: HoneyPotStore;

	constructor() {
		this.router = new RouterStore<RootStore>(this);
		this.wallet = new WalletStore(this);
		this.contracts = new ContractsStore(this);
		this.airdrops = new AirdropStore(this);
		this.rebase = new RebaseStore(this);
		this.rewards = new RewardsStore(this);
		this.uiState = new UiState(this);
		this.ibBTCStore = new IbBTCStore(this);
		this.honeyPot = new HoneyPotStore(this);
		// RenVM transactions store.
		this.transactions = new TransactionsStore(this);
		this.setts = new SettStoreV2(this);

		this.walletRefresh();
	}

	async walletRefresh(): Promise<void> {
		const chain = this.wallet.network.name;
		const refreshData = [
			this.setts.loadAssets(chain),
			this.setts.loadPrices(chain),
			this.wallet.getGasPrice(),
			this.contracts.updateProvider(),
		];
		if (chain === NETWORK_LIST.ETH) {
			refreshData.push(this.setts.loadGeysers(chain));
			refreshData.push(this.rebase.fetchRebaseStats());
		} else {
			refreshData.push(this.setts.loadSetts(chain));
		}
		await Promise.all(refreshData);

		if (this.wallet.connectedAddress) {
			this.contracts.updateProvider();
			await this.wallet.getGasPrice();
			await this.contracts.fetchContracts();
			if (chain === NETWORK_LIST.ETH) {
				this.uiState.reduceRebase();
				this.ibBTCStore.init();
				this.rewards.fetchSettRewards();
				this.uiState.reduceTreeRewards();
				this.airdrops.fetchAirdrops();
				this.uiState.reduceAirdrops();
			}
			this.uiState.reduceStats();
		}
	}
}

const store = new RootStore();

export default store;
