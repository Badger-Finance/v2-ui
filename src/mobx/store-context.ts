import { RootStore } from './store';
import { createContext } from 'react';

// Use React context to make your store available in your application
export const StoreContext = createContext({} as RootStore);
export const StoreProvider = StoreContext.Provider;
