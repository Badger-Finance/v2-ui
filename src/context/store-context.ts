import { createContext } from "react";
import { RootStore } from "../mobx/store";

// Use React context to make your store available in your application
export const StoreContext = createContext({} as RootStore);
export const StoreProvider = StoreContext.Provider;