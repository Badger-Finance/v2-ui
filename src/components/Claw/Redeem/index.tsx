import React from 'react';
import { Box, Grid, InputBase, makeStyles, Typography } from '@material-ui/core';
import { StoreContext } from 'mobx/store-context';
import BigNumber from 'bignumber.js';
import { ethers } from 'ethers';
import dayjs from 'dayjs';

import { TokenSelectorLabel } from 'components-v2/common/TokenSelectorLabel';
import { TokenSelectorWithAmountContainer } from 'components-v2/common/TokenSelectorWithAmountContainer';
import { TokenSelect } from 'components-v2/common/TokenSelect';
import { TokenAmountInput } from 'components-v2/common/TokenAmountInput';
import { PercentageGroup } from 'components-v2/common/PercentageGroup';
import { ClawDetails } from '../ClawDetails';
import { ActionButton } from '../ActionButton';
import { scaleToString, Direction } from 'utils/componentHelpers';
import { validateAmountBoundaries } from 'utils/componentHelpers';
import { useAmountToReceive, useDetails, useError } from './redeem.hooks';
import { ClawParam } from '../claw.model';
import { observer } from 'mobx-react-lite';

const useStyles = makeStyles((theme) => ({
	border: {
		border: '1px solid #5C5C5C',
		borderRadius: 8,
	},
	selectContainer: {
		[theme.breakpoints.only('xs')]: {
			justifyContent: 'space-between',
		},
		[theme.breakpoints.up('lg')]: {
			paddingLeft: '10%',
		},
	},
	margin: {
		margin: theme.spacing(1),
	},
	centered: {
		margin: 'auto',
	},
}));

const Redeem = observer(() => {
	const { claw: store, contracts, wallet } = React.useContext(StoreContext);
	const { collaterals, claws, syntheticsDataByEMP, sponsorInformationByEMP } = store;
	const classes = useStyles();
	const [redeem, setRedeemParams] = React.useState<ClawParam>({});
	const details = useDetails(redeem);
	const error = useError(redeem);

	const { selectedOption, amount } = redeem;
	const selectedSynthetic = syntheticsDataByEMP.get(selectedOption || '');
	const bToken = contracts.tokens[selectedSynthetic?.collateralCurrency ?? ''];
	const decimals = bToken ? bToken.decimals : 18; // Default to 18 decimals.
	const clawBalance = sponsorInformationByEMP.get(selectedOption || '')?.position.tokensOutstanding;
	const amountToReceive = useAmountToReceive(redeem, decimals);
	const isSyntheticExpired =
		selectedSynthetic && dayjs(selectedSynthetic.expirationTimestamp.toNumber() * 1000).isBefore(dayjs().utc());
	const actionText = isSyntheticExpired ? 'SETTLE' : 'REDEEM';

	const handleRedeem = async () => {
		const [empAddress, numTokens] = [selectedOption, amount];
		if (!empAddress || !numTokens || !selectedSynthetic) return;

		const isSyntheticExpired = dayjs(selectedSynthetic.expirationTimestamp.toNumber() * 1000).isBefore(
			dayjs().utc(),
		);

		// we settle expire if the synthetic is expired
		if (isSyntheticExpired) {
			await store.actionStore.settleExpired(
				empAddress,
				ethers.utils.parseUnits(numTokens, decimals).toHexString(),
			);
		} else {
			await store.actionStore.redeem(empAddress, ethers.utils.parseUnits(numTokens, decimals).toHexString());
		}
	};

	return (
		<Grid container>
			<Box clone pb={2}>
				<Grid item xs={12}>
					<TokenSelectorWithAmountContainer
						tokenBalanceInformation={
							<TokenSelectorLabel
								name="Token"
								balanceLabel={selectedOption && `Available ${claws.get(selectedOption)}:`}
								balance={selectedOption && scaleToString(clawBalance, decimals, Direction.Down)}
							/>
						}
						tokenList={
							<TokenSelect
								placeholder="Select Token"
								selectedOption={selectedOption}
								options={claws}
								disabled={!wallet.connectedAddress || claws.size === 0}
								onChange={(selectedOption: string) => {
									setRedeemParams({
										selectedOption,
										amount: undefined,
										error: undefined,
									});
								}}
							/>
						}
						tokenAmount={
							<TokenAmountInput
								value={amount}
								disabled={!selectedOption || !wallet.connectedAddress}
								onChange={(amount: string) => {
									if (!clawBalance || !bToken) return;
									setRedeemParams({
										selectedOption,
										amount,
										error: validateAmountBoundaries({ amount, maximum: clawBalance }),
									});
								}}
							/>
						}
						percentagesGroup={
							<PercentageGroup
								disabled={!selectedOption || !wallet.connectedAddress}
								options={[25, 50, 75, 100]}
								onChange={(percentage: number) => {
									if (!clawBalance || !bToken) return;
									setRedeemParams({
										selectedOption,
										amount: clawBalance
											.multipliedBy(percentage / 100)
											.dividedBy(10 ** decimals)
											.toFixed(decimals, BigNumber.ROUND_DOWN),
										error: undefined,
									});
								}}
							/>
						}
					/>
				</Grid>
			</Box>
			{bToken && (
				<Box clone py={2}>
					<Grid item xs={12} sm={8} className={classes.centered}>
						<Box clone pb={1}>
							<Grid item xs={12}>
								<TokenSelectorLabel name="You Receive" />
							</Grid>
						</Box>
						<Box clone py={1} px={2}>
							<Grid container alignContent="center" alignItems="center" className={classes.border}>
								<Grid item xs={12}>
									<Grid container alignItems="center" spacing={2} className={classes.selectContainer}>
										<Grid item xs={12} sm={6}>
											<Typography>{collaterals.get(bToken.address)}</Typography>
										</Grid>
										<Grid item xs={12} sm={6}>
											<InputBase
												type="tel"
												disabled
												placeholder="0.00"
												value={scaleToString(amountToReceive, decimals, Direction.Down)}
											/>
										</Grid>
									</Grid>
								</Grid>
							</Grid>
						</Box>
					</Grid>
				</Box>
			)}
			<Grid item xs={12}>
				<ClawDetails details={details} />
			</Grid>
			<Grid item xs={12}>
				<Grid container>
					<ActionButton text={error ? error : actionText} disabled={!!error} onClick={handleRedeem} />
				</Grid>
			</Grid>
		</Grid>
	);
});

export default Redeem;
