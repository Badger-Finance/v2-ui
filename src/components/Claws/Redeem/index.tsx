import React from 'react';
import { Box, Grid, InputBase, makeStyles, Typography } from '@material-ui/core';
import { StoreContext } from 'mobx/store-context';
import BigNumber from 'bignumber.js';
import { ethers } from 'ethers';

import { TokenAmountLabel } from 'components-v2/common/TokenAmountLabel';
import { TokenAmountSelector } from 'components-v2/common/TokenAmountSelector';
import { ClawDetails } from '../ClawDetails';
import { ActionButton } from '../ActionButton';
import { scaleToString, Direction } from 'utils/componentHelpers';
import { validateAmountBoundaries } from 'utils/componentHelpers';
import { useMainStyles } from '../index';
import { useAmountToReceive, useDetails, useError } from './redeem.hooks';
import { ClawParam } from '../claw-param.model';
import dayjs from 'dayjs';

const useStyles = makeStyles((theme) => ({
	border: {
		border: '1px solid #5C5C5C',
		borderRadius: 8,
	},
	selectContainer: {
		[theme.breakpoints.only('xs')]: {
			justifyContent: 'space-between',
		},
		[theme.breakpoints.up('lg')]: {
			paddingLeft: '10%',
		},
	},
	margin: {
		margin: theme.spacing(1),
	},
	centered: {
		margin: 'auto',
	},
}));

const Redeem = () => {
	const { claw: store, contracts, wallet } = React.useContext(StoreContext);
	const { collaterals, claws, syntheticsDataByEMP, sponsorInformationByEMP } = store;
	const mainClasses = useMainStyles();
	const classes = useStyles();
	const [redeem, setRedeemParams] = React.useState<ClawParam>({});
	const details = useDetails(redeem);
	const error = useError(redeem);

	const { selectedOption, amount } = redeem;
	const selectedSynthetic = syntheticsDataByEMP.get(selectedOption || '');
	const bToken = contracts.tokens[selectedSynthetic?.collateralCurrency ?? ''];
	const decimals = bToken ? bToken.decimals : 18; // Default to 18 decimals.
	const clawBalance = sponsorInformationByEMP.get(selectedOption || '')?.position.tokensOutstanding;
	const amountToReceive = useAmountToReceive(redeem, decimals);
	const isSyntheticExpired =
		selectedSynthetic && dayjs(selectedSynthetic.expirationTimestamp.toNumber() * 1000).isBefore(dayjs().utc());
	const actionText = isSyntheticExpired ? 'SETTLE' : 'REDEEM';

	const handleRedeem = () => {
		const [empAddress, numTokens] = [selectedOption, amount];
		if (!empAddress || !numTokens || !selectedSynthetic) return;

		const isSyntheticExpired = dayjs(selectedSynthetic.expirationTimestamp.toNumber() * 1000).isBefore(
			dayjs().utc(),
		);

		// we settle expire if the synthetic is expired
		if (isSyntheticExpired) {
			store.actionStore.settleExpired(empAddress, ethers.utils.parseUnits(numTokens, decimals).toHexString());
		} else {
			store.actionStore.redeem(empAddress, ethers.utils.parseUnits(numTokens, decimals).toHexString());
		}
	};

	return (
		<Grid container>
			<Box clone pb={2}>
				<Grid item xs={12}>
					<Box clone pb={1}>
						<Grid item xs={12}>
							<TokenAmountLabel
								name="Token"
								balanceLabel={selectedOption && `Available ${claws.get(selectedOption)}:`}
								balance={selectedOption && scaleToString(clawBalance, decimals, Direction.Down)}
							/>
						</Grid>
					</Box>
					<TokenAmountSelector
						placeholder="Select Token"
						options={claws}
						selectedOption={selectedOption}
						displayAmount={amount}
						disabledOptions={!wallet.connectedAddress || claws.size === 0}
						disabledAmount={!selectedOption || !wallet.connectedAddress}
						onAmountChange={(amount: string) => {
							if (!clawBalance || !bToken) return;
							setRedeemParams({
								selectedOption,
								amount,
								error: validateAmountBoundaries({ amount, maximum: clawBalance }),
							});
						}}
						onOptionChange={(selectedOption: string) => {
							setRedeemParams({
								selectedOption,
								amount: undefined,
								error: undefined,
							});
						}}
						onApplyPercentage={(percentage) => {
							if (!clawBalance || !bToken) return;
							setRedeemParams({
								selectedOption,
								amount: clawBalance
									.multipliedBy(percentage / 100)
									.dividedBy(10 ** decimals)
									.toFixed(0, BigNumber.ROUND_DOWN),
								error: undefined,
							});
						}}
					/>
				</Grid>
			</Box>
			{bToken && (
				<Box clone py={2}>
					<Grid item xs={12} sm={8} className={classes.centered}>
						<Box clone pb={1}>
							<Grid item xs={12}>
								<TokenAmountLabel name="You Receive" />
							</Grid>
						</Box>
						<Box clone py={1} px={2}>
							<Grid container alignContent="center" alignItems="center" className={classes.border}>
								<Grid item xs={12}>
									<Grid container alignItems="center" spacing={2} className={classes.selectContainer}>
										<Grid item xs={12} sm={6}>
											<Typography>{collaterals.get(bToken.address)}</Typography>
										</Grid>
										<Grid item xs={12} sm={6}>
											<InputBase
												type="tel"
												disabled
												placeholder="0.00"
												value={scaleToString(amountToReceive, decimals, Direction.Down)}
											/>
										</Grid>
									</Grid>
								</Grid>
							</Grid>
						</Box>
					</Grid>
				</Box>
			)}
			<Grid item xs={12}>
				<Grid container className={mainClasses.details}>
					<ClawDetails details={details} />
				</Grid>
			</Grid>
			<Grid item xs={12}>
				<Grid container>
					<ActionButton text={error ? error : actionText} disabled={!!error} onClick={handleRedeem} />
				</Grid>
			</Grid>
		</Grid>
	);
};

export default Redeem;
