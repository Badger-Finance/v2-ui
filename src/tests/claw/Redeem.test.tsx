import React from 'react';
import '@testing-library/jest-dom';
import store from '../../mobx/store';
import {
	mockClaws,
	mockCollaterals,
	mockSponsorInformationByEmp,
	mockSyntheticDataByEmp,
	mockTokens,
} from './claw.mocks';
import { customRender, fireEvent, screen, within } from '../Utils';
import { StoreProvider } from '../../mobx/store-context';
import Redeem from '../../components/Claws/Redeem';
import * as RedeemHooks from '../../components/Claws/Redeem/redeem.hooks';

describe('Claw - Redeem', () => {
	beforeAll(() => {
		store.wallet.connectedAddress = '0x1a1a1a1a1a1a1a1a1a1a1a1a1a1a1a1a1a1a1a1a';
		store.contracts.tokens = mockTokens;
		store.claw.collaterals = mockCollaterals;
		store.claw.syntheticsDataByEMP = mockSyntheticDataByEmp;
		store.claw.sponsorInformationByEMP = mockSponsorInformationByEmp;
		store.claw.claws = mockClaws;
	});

	beforeEach(() => {
		customRender(
			<StoreProvider value={store}>
				<Redeem />
			</StoreProvider>,
		);
	});

	test('matches snapshot', () => {
		const { container } = customRender(
			<StoreProvider value={store}>
				<Redeem />
			</StoreProvider>,
		);
		expect(container).toMatchSnapshot();
	});

	test('token list has synthetics information', () => {
		const tokenSelector = screen.getByRole('button', { name: 'Select Token' });
		expect(tokenSelector).toBeEnabled();
		fireEvent.mouseDown(tokenSelector);
		const tokenList = within(screen.getByRole('listbox'));
		expect(tokenList.getByText('USD/bBadger 5-29')).toBeInTheDocument();
		expect(tokenList.getByText('USD-[bwBTC/ETH SLP] 5-29')).toBeInTheDocument();
	});

	describe('when no synthetic token is selected', () => {
		test('percentage group is disabled', () => {
			expect(
				screen.getByRole('button', {
					name: '50%',
				}),
			).toBeDisabled();
		});

		test('action button is disabled and with text "Select a Token"', () => {
			expect(screen.getByRole('button', { name: 'Select a Token' })).toBeDisabled();
		});
	});

	describe('when a synthetic token is selected', () => {
		beforeEach(() => {
			fireEvent.mouseDown(screen.getByRole('button', { name: 'Select Token' }));
			fireEvent.click(within(screen.getByRole('listbox')).getByText('USD/bBadger 5-29'));
		});

		test('synthetic token balance is displayed', () => {
			expect(screen.getByText('Available USD/bBadger 5-29:')).toBeInTheDocument();
			expect(screen.getByText('44177.984375000041211331')).toBeInTheDocument();
		});

		test('action button is disabled and with text "Enter an amount"', () => {
			expect(screen.getByRole('button', { name: 'Enter an amount' })).toBeDisabled();
		});

		test('synthetic amount can be changed', () => {
			const [amountInput] = screen.getAllByRole('textbox');
			fireEvent.change(amountInput, { target: { value: '123' } });
			expect(screen.getByDisplayValue('123')).toBeInTheDocument();
		});

		test('synthetic token balance percentage can applied', () => {
			fireEvent.click(
				screen.getByRole('button', {
					name: '50%',
				}),
			);
			expect(screen.getByDisplayValue('22088.992187500020605665'));
		});
	});

	describe('when synthetic amount is inputted', () => {
		beforeEach(() => {
			jest.clearAllMocks();
			jest.spyOn(RedeemHooks, 'useDetails').mockReturnValue([
				{ name: 'Expiration Date', text: 'May 29, 2021 18:00 UTC' },
				{ name: 'Expiration Price', text: '-' },
			]);
			fireEvent.mouseDown(screen.getByRole('button', { name: 'Select Token' }));
			fireEvent.click(within(screen.getByRole('listbox')).getByText('USD/bBadger 5-29'));
			const [amountInput] = screen.getAllByRole('textbox');
			fireEvent.change(amountInput, { target: { value: '123' } });
		});

		test('collateral balance to receive is displayed', () => {
			expect(screen.getByText('bBADGER')).toBeInTheDocument();
			screen.debug(undefined, 1e18);
			expect(screen.getByDisplayValue('0.15107035573335676819')).toBeInTheDocument();
		});

		test('redeem details are displayed', () => {
			expect(screen.getByText('May 29, 2021 18:00 UTC')).toBeInTheDocument();
		});

		test('action button gets enabled', () => {
			expect(screen.getByRole('button', { name: 'REDEEM' })).toBeEnabled();
		});
	});
});
