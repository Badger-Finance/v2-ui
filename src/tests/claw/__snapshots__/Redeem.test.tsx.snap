// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Claw - Redeem matches snapshot 1`] = `
<div>
  <div
    class="MuiGrid-root-5 MuiGrid-container-6"
  >
    <div
      class="MuiGrid-root-5 MuiBox-root-108 MuiBox-root-1 MuiGrid-item-7 MuiGrid-grid-xs-12-51"
    >
      <div
        class="MuiGrid-root-5 MuiGrid-container-6"
      >
        <div
          class="MuiGrid-root-5 MuiBox-root-108 MuiBox-root-2 MuiGrid-item-7 MuiGrid-grid-xs-12-51"
        >
          <div
            class="MuiGrid-root-5 MuiGrid-container-6 MuiGrid-spacing-xs-1-28"
          >
            <div
              class="MuiGrid-root-5 MuiGrid-item-7 MuiGrid-grid-xs-12-51 MuiGrid-grid-sm-6-59"
            >
              <p
                class="MuiTypography-root-115 MuiTypography-body2-116 MuiTypography-colorTextSecondary-141"
              >
                Token
              </p>
            </div>
            <div
              class="MuiGrid-root-5 MuiGrid-item-7 MuiGrid-grid-xs-12-51 MuiGrid-grid-sm-6-59"
            >
              <div
                class="MuiGrid-root-5 makeStyles-availableText-113 MuiGrid-container-6 MuiGrid-spacing-xs-1-28"
              >
                <div
                  class="MuiGrid-root-5 MuiGrid-item-7"
                >
                  <p
                    class="MuiTypography-root-115 MuiTypography-body2-116 MuiTypography-colorTextSecondary-141"
                  />
                </div>
                <div
                  class="MuiGrid-root-5 makeStyles-balanceText-114 MuiGrid-item-7"
                >
                  <p
                    class="MuiTypography-root-115 MuiTypography-body2-116 MuiTypography-colorTextPrimary-140"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root-5 MuiGrid-item-7 MuiGrid-grid-xs-12-51"
        >
          <div
            class="MuiGrid-root-5 makeStyles-border-110 MuiBox-root-108 MuiBox-root-3 MuiGrid-container-6 MuiGrid-align-items-xs-center-14 MuiGrid-align-content-xs-center-18"
          >
            <div
              class="MuiGrid-root-5 MuiGrid-item-7 MuiGrid-grid-xs-12-51 MuiGrid-grid-sm-8-61 MuiGrid-grid-lg-9-90"
            >
              <div
                class="MuiGrid-root-5 makeStyles-selectContainer-111 MuiGrid-container-6 MuiGrid-spacing-xs-2-29 MuiGrid-align-items-xs-center-14"
              >
                <div
                  class="MuiGrid-root-5 MuiGrid-item-7 MuiGrid-grid-xs-12-51 MuiGrid-grid-sm-true-53"
                >
                  <div
                    class="MuiFormControl-root-147 makeStyles-margin-146"
                  >
                    <div
                      class="MuiInputBase-root-162 MuiInputBase-formControl-163"
                    >
                      <div
                        aria-haspopup="listbox"
                        class="MuiSelect-root-151 MuiSelect-select-152 MuiSelect-selectMenu-155 MuiInputBase-input-173"
                        role="button"
                        tabindex="0"
                      >
                        Select Token
                      </div>
                      <input
                        aria-hidden="true"
                        class="MuiSelect-nativeInput-161"
                        placeholder="Select Token"
                        tabindex="-1"
                        value=""
                      />
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root-180 MuiSelect-icon-157"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M7 10l5 5 5-5z"
                        />
                      </svg>
                    </div>
                  </div>
                </div>
                <div
                  class="MuiGrid-root-5 MuiGrid-item-7 MuiGrid-grid-xs-12-51 MuiGrid-grid-sm-true-53"
                >
                  <div
                    class="MuiInputBase-root-162 makeStyles-input-193 MuiInputBase-disabled-165 MuiInputBase-fullWidth-172"
                  >
                    <input
                      class="MuiInputBase-input-173 MuiInputBase-disabled-165"
                      disabled=""
                      pattern="^[0-9]*[.,]?[0-9]*$"
                      placeholder="0.00"
                      type="tel"
                      value=""
                    />
                  </div>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root-5 MuiGrid-item-7 MuiGrid-grid-xs-12-51 MuiGrid-grid-sm-4-57 MuiGrid-grid-lg-3-84"
            >
              <div
                class="MuiGrid-root-5 MuiGrid-container-6 MuiGrid-justify-xs-flex-end-24"
              >
                <div
                  aria-label="text button group"
                  class="MuiButtonGroup-root-195"
                  role="group"
                >
                  <button
                    class="MuiButtonBase-root-248 MuiButton-root-219 MuiButton-text-221 MuiButtonGroup-grouped-201 MuiButtonGroup-groupedHorizontal-202 MuiButtonGroup-groupedText-204 MuiButtonGroup-groupedTextHorizontal-205 MuiButtonGroup-groupedText-204 MuiButtonGroup-disabled-198 makeStyles-button-194 MuiButton-textSizeSmall-234 MuiButton-sizeSmall-240 MuiButton-disableElevation-230 MuiButton-disabled-232 MuiButtonBase-disabled-249"
                    disabled=""
                    tabindex="-1"
                    type="button"
                  >
                    <span
                      class="MuiButton-label-220"
                    >
                      25
                      %
                    </span>
                  </button>
                  <button
                    class="MuiButtonBase-root-248 MuiButton-root-219 MuiButton-text-221 MuiButtonGroup-grouped-201 MuiButtonGroup-groupedHorizontal-202 MuiButtonGroup-groupedText-204 MuiButtonGroup-groupedTextHorizontal-205 MuiButtonGroup-groupedText-204 MuiButtonGroup-disabled-198 makeStyles-button-194 MuiButton-textSizeSmall-234 MuiButton-sizeSmall-240 MuiButton-disableElevation-230 MuiButton-disabled-232 MuiButtonBase-disabled-249"
                    disabled=""
                    tabindex="-1"
                    type="button"
                  >
                    <span
                      class="MuiButton-label-220"
                    >
                      50
                      %
                    </span>
                  </button>
                  <button
                    class="MuiButtonBase-root-248 MuiButton-root-219 MuiButton-text-221 MuiButtonGroup-grouped-201 MuiButtonGroup-groupedHorizontal-202 MuiButtonGroup-groupedText-204 MuiButtonGroup-groupedTextHorizontal-205 MuiButtonGroup-groupedText-204 MuiButtonGroup-disabled-198 makeStyles-button-194 MuiButton-textSizeSmall-234 MuiButton-sizeSmall-240 MuiButton-disableElevation-230 MuiButton-disabled-232 MuiButtonBase-disabled-249"
                    disabled=""
                    tabindex="-1"
                    type="button"
                  >
                    <span
                      class="MuiButton-label-220"
                    >
                      75
                      %
                    </span>
                  </button>
                  <button
                    class="MuiButtonBase-root-248 MuiButton-root-219 MuiButton-text-221 MuiButtonGroup-grouped-201 MuiButtonGroup-groupedHorizontal-202 MuiButtonGroup-groupedText-204 MuiButtonGroup-groupedTextHorizontal-205 MuiButtonGroup-groupedText-204 MuiButtonGroup-disabled-198 makeStyles-button-194 MuiButton-textSizeSmall-234 MuiButton-sizeSmall-240 MuiButton-disableElevation-230 MuiButton-disabled-232 MuiButtonBase-disabled-249"
                    disabled=""
                    tabindex="-1"
                    type="button"
                  >
                    <span
                      class="MuiButton-label-220"
                    >
                      100
                      %
                    </span>
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root-5 MuiGrid-item-7 MuiGrid-grid-xs-12-51"
    >
      <div
        class="MuiGrid-root-5 makeStyles-root-251 MuiGrid-container-6"
      >
        <div
          class="MuiGrid-root-5 makeStyles-row-252 MuiGrid-container-6"
        >
          <div
            class="MuiGrid-root-5 makeStyles-detailName-253 MuiGrid-item-7 MuiGrid-grid-xs-6-45"
          >
            <p
              class="MuiTypography-root-115 MuiTypography-body2-116 MuiTypography-colorTextSecondary-141"
            >
              Expiration Date
            </p>
          </div>
          <div
            class="MuiGrid-root-5 makeStyles-placeholder-255 MuiGrid-container-6 MuiGrid-item-7 MuiGrid-grid-xs-6-45"
          >
            <div
              class="MuiGrid-root-5 MuiGrid-item-7 MuiGrid-grid-xs-12-51"
            >
              <p
                class="MuiTypography-root-115 MuiTypography-body2-116 MuiTypography-colorTextPrimary-140"
              >
                -
              </p>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root-5 makeStyles-row-252 MuiGrid-container-6"
        >
          <div
            class="MuiGrid-root-5 makeStyles-detailName-253 MuiGrid-item-7 MuiGrid-grid-xs-6-45"
          >
            <p
              class="MuiTypography-root-115 MuiTypography-body2-116 MuiTypography-colorTextSecondary-141"
            >
              Expiration Price
            </p>
          </div>
          <div
            class="MuiGrid-root-5 makeStyles-placeholder-255 MuiGrid-container-6 MuiGrid-item-7 MuiGrid-grid-xs-6-45"
          >
            <div
              class="MuiGrid-root-5 MuiGrid-item-7 MuiGrid-grid-xs-12-51"
            >
              <p
                class="MuiTypography-root-115 MuiTypography-body2-116 MuiTypography-colorTextPrimary-140"
              >
                -
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root-5 MuiGrid-item-7 MuiGrid-grid-xs-12-51"
    >
      <div
        class="MuiGrid-root-5 MuiGrid-container-6"
      >
        <button
          class="MuiButtonBase-root-248 MuiButton-root-219 MuiButton-contained-227 makeStyles-button-256 MuiButton-containedPrimary-228 MuiButton-containedSizeLarge-239 MuiButton-sizeLarge-241 MuiButton-disabled-232 MuiButtonBase-disabled-249"
          disabled=""
          tabindex="-1"
          type="button"
        >
          <span
            class="MuiButton-label-220"
          >
            Select a Token
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
